### ISSUE TITLE
Support kernel tracing in drcachesim for system calls

### ISSUE THREAD

**Issue by abhinav92003** [DynamoRIO/dynamorio#5505]
This issue is to track the design and implementation for the support to trace kernel instructions and data addresses for system calls in  drcachesim.

The current plan is to use Intel-PT to collect system call instruction traces in the drcachesim tracer. We can start kernel tracing by using perf_events_open + ioctl in a pre-syscall event, and end it using ioctl in a post-syscall event.

We can then convert the PT trace to a drcachesim trace using libipt, and interleave it into the user-mode trace gathered by drcachesim, in raw2trace.

An approximation of the data addresses accessed by the kernel can be reconstructed using some heuristics and metadata collected by drcachesim (e.g. system call parameters). There are some interesting research articles that implement similar techniques: https://ieeexplore.ieee.org/document/9065578.

**Comment by prasun3** [comment]
Purely software based approach (like some prior academic work) not being considered?

https://sorav.compiler.ai/pubs/btkernel.pdf
http://www.cs.toronto.edu/~peter/feiner_asplos_2012.pdf

**Comment by abhinav92003** [comment]
We do not want to assume the capability of modifying the kernel or loading a kernel module at runtime.

**Comment by algr** [comment]
It ought to be possible to do this on Arm as well using cs_etm events; the equivalent decoder library is https://github.com/Linaro/OpenCSD. Would it be possible for this to be done in a way where the main logic is cross-architecture, and only the bits that are specific to PT (resp ETM) are in the Intel (resp Arm) parts?

**Comment by abhinav92003** [comment]
Yes, we do plan to use the same hooks in the drmemtrace tracer and raw2trace that we use today for Intel-PT traces to also support ETM traces. We're blocked on getting access to required hardware with ETM capability though.

**Comment by rejohn** [comment]
Is this feature fully supported now?

**Comment by abhinav92003** [comment]
The basic support for collecting Intel-PT traces during system calls is present. We also have some unit tests but they're not run in our Github Actions test suite because they need a machine with Intel-PT support. You can run them manually on a suitable machine by following the instructions at https://dynamorio.org/page_drpttracer.html#sec_unit_tests.

There are still some missing pieces and rough edges in this feature though:
- PT traces are instruction-only. Our plan was to reconstruct an address stream that preserves some useful properties of the original (e.g., the stack distance profile), much like DWT from HPCA 2020 (https://ieeexplore.ieee.org/document/9065578).
- Some optimizations towards making this feature usable on large workloads are still missing, e.g., streaming libipt decoding support, removal of noise instructions.
https://github.com/DynamoRIO/dynamorio/pull/6179 unfortunately isn't merged yet.

We welcome contributions from the community so I'm happy to discuss more details.

**Comment by rejohn** [comment]
I'm currently researching instruction stream analysis based on ARM CoreSight, which similarly only supports instruction tracing. I don't yet have a good idea for reconstructing the data stream,  If you already have a good implementation idea, please share it with me. I would be very grateful.

**Comment by rejohn** [comment]
I saw in Dynamorio-User that you are researching a scheme for ”synthesizing“ memory addresses. in a way that preserves usefulness of the memory access pattern,  I wonder if this scheme is about to be phased out?
